// permissionless/_cjs/actions/erc7579/uninstallModules.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.uninstallModules = void 0;
const viem_1 = require("viem");
const account_abstraction_1 = require("viem/account-abstraction");
const utils_1 = require("viem/utils");
const utils_2 = require("viem/utils");
const errors_1 = require("../../errors/index.js");
const supportsModule_1 = require("./supportsModule.js");
async function uninstallModules(client, parameters) {
    const { account: account_ = client.account, maxFeePerGas, maxPriorityFeePerGas, nonce, modules } = parameters;
    if (!account_) {
        throw new errors_1.AccountNotFoundError({
            docsPath: "/docs/actions/wallet/sendTransaction"
        });
    }
    const account = (0, utils_2.parseAccount)(account_);
    return (0, utils_1.getAction)(client, account_abstraction_1.sendUserOperation, "sendUserOperation")({
        calls: modules.map(({ type, address, context }) => ({
            to: account.address,
            value: BigInt(0),
            data: (0, viem_1.encodeFunctionData)({
                abi: [
                    {
                        name: "uninstallModule",
                        type: "function",
                        stateMutability: "nonpayable",
                        inputs: [
                            {
                                type: "uint256",
                                name: "moduleTypeId"
                            },
                            {
                                type: "address",
                                name: "module"
                            },
                            {
                                type: "bytes",
                                name: "deInitData"
                            }
                        ],
                        outputs: []
                    }
                ],
                functionName: "uninstallModule",
                args: [(0, supportsModule_1.parseModuleTypeId)(type), (0, viem_1.getAddress)(address), context]
            })
        })),
        maxFeePerGas,
        maxPriorityFeePerGas,
        nonce,
        account
    });
}
exports.uninstallModules = uninstallModules;
