// permissionless/_cjs/utils/encode7579Calls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.encode7579Calls = void 0;
const viem_1 = require("viem");
const supportsExecutionMode_1 = require("../actions/erc7579/supportsExecutionMode.js");
function encode7579Calls({ mode, callData }) {
    if (callData.length > 1 && mode?.type !== "batchcall") {
        throw new Error(`mode ${JSON.stringify(mode)} does not supported for batchcall calldata`);
    }
    const executeAbi = [
        {
            type: "function",
            name: "execute",
            inputs: [
                {
                    name: "execMode",
                    type: "bytes32",
                    internalType: "ExecMode"
                },
                {
                    name: "executionCalldata",
                    type: "bytes",
                    internalType: "bytes"
                }
            ],
            outputs: [],
            stateMutability: "payable"
        }
    ];
    if (callData.length > 1) {
        return (0, viem_1.encodeFunctionData)({
            abi: executeAbi,
            functionName: "execute",
            args: [
                (0, supportsExecutionMode_1.encodeExecutionMode)(mode),
                (0, viem_1.encodeAbiParameters)([
                    {
                        name: "executionBatch",
                        type: "tuple[]",
                        components: [
                            {
                                name: "target",
                                type: "address"
                            },
                            {
                                name: "value",
                                type: "uint256"
                            },
                            {
                                name: "callData",
                                type: "bytes"
                            }
                        ]
                    }
                ], [
                    callData.map((arg) => {
                        return {
                            target: arg.to,
                            value: arg.value ?? 0n,
                            callData: arg.data ?? "0x"
                        };
                    })
                ])
            ]
        });
    }
    return (0, viem_1.encodeFunctionData)({
        abi: executeAbi,
        functionName: "execute",
        args: [
            (0, supportsExecutionMode_1.encodeExecutionMode)(mode),
            (0, viem_1.concatHex)([
                callData[0].to,
                (0, viem_1.toHex)(callData[0].value ?? 0n, { size: 32 }),
                callData[0].data ?? "0x"
            ])
        ]
    });
}
exports.encode7579Calls = encode7579Calls;
