// permissionless/_cjs/accounts/kernel/utils/signTypedData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.signTypedData = void 0;
const viem_1 = require("viem");
const isKernelV2_1 = require("./isKernelV2.js");
const wrapMessageHash_1 = require("./wrapMessageHash.js");
async function signTypedData(parameters) {
    const { owner, accountAddress, kernelVersion: accountVersion, chainId, ...typedData } = parameters;
    if ((0, isKernelV2_1.isKernelV2)(accountVersion)) {
        return owner.signTypedData({
            ...typedData
        });
    }
    const { message, primaryType, types: _types, domain } = typedData;
    const types = {
        EIP712Domain: (0, viem_1.getTypesForEIP712Domain)({
            domain: domain
        }),
        ..._types
    };
    (0, viem_1.validateTypedData)({
        domain,
        message,
        primaryType,
        types
    });
    const typedHash = (0, viem_1.hashTypedData)({ message, primaryType, types, domain });
    const wrappedMessageHash = (0, wrapMessageHash_1.wrapMessageHash)(typedHash, {
        kernelVersion: accountVersion,
        accountAddress,
        chainId: chainId
    });
    return owner.signMessage({
        message: { raw: wrappedMessageHash }
    });
}
exports.signTypedData = signTypedData;
